sandbox/grist/predicate_formula.py

```diff
@@ -63,6 +65,60 @@ def parse_predicate_formula_json(formula):
   'None': None,
 }
 
+
+def process_renames(formula, collector, renamer):
+
+  """
+  Given a predicate formula, a collector and a renamer, rename all references in the formula
+  that the renamer wants to rename. This is used to automatically update references in an ACL
+  or dropdown condition formula when a column it refers to has been renamed.
+
+  The collector should be a subclass of TreeConverter that collects related NamedEntity's and
+  stores them in the field "entities". See acl._ACLEntityCollector for an example.
+
+  The renamer should be a function taking a NamedEntity as its only argument. It should return 
+  a new name for this NamedEntity when it wants to rename this entity, or None otherwise.
+  """
+  patches = []
+  # "$" can be used to refer to "rec." in Grist formulas, but it is not valid Python.
+  # We need to replace it with "rec." before parsing the formula, and restore it back after
+  # the surgery.
+  # Keep the dollar replacer object, so that later we know how to restore properly.
+  dollar_replacer = get_dollar_replacer(formula)
+  formula_nodollar = dollar_replacer.get_text()
+  try:
+    atok = asttokens.ASTTokens(formula_nodollar, tree=ast.parse(formula_nodollar, mode='eval'))
+    collector.visit(atok.tree)
+  except SyntaxError:
+    # Don't do anything to a syntactically wrong formula.
+    return formula
+  except ValueError as e:
```

Comment:
I tried doing that initially. I remember giving up because the `ValueError` was raised by `TreeConverter`, which I thought might be used elsewhere.

A revisit of this issue suggests it might actually be good to raise `SyntaxError` instead, as most code using `TreeConverter` has been merged, and `predicate_formula.py` has become its sole user.